src/bridge/test-peer.c

```diff
@@ -470,6 +483,110 @@ test_reopen (TestCase *tc,
   g_free (bridge);
 }
 
+static void
+test_timeout (TestCase *tc,
+             gconstpointer unused)
+{
+  gchar *bridge;
+  GBytes *sent;
+  JsonObject *control;
+  CockpitTransport *other = NULL;
+  gboolean closed = FALSE;
+
+  bridge = g_strdup_printf("{ \"match\": { \"payload\": \"upper\" }, \"timeout\": 1, \"spawn\": [ \"/%s\", \"--upper\", \"--count\" ] }", BUILDDIR "/mock-bridge");
+  tc->peer = peer_new (tc->transport, bridge);
+
+  emit_string (tc, NULL, "{\"command\": \"open\", \"channel\": \"a\", \"payload\": \"upper\"}");
+  emit_string (tc, "a", "Oh MarmaLade");
+
+  while ((control = mock_transport_pop_control (tc->transport)) == NULL)
+    g_main_context_iteration (NULL, TRUE);
+  cockpit_assert_json_eq (control, "{\"command\":\"ready\",\"channel\":\"a\",\"count\":0}");
+  control = NULL;
+
+  while ((sent = mock_transport_pop_channel (tc->transport, "a")) == NULL)
+    g_main_context_iteration (NULL, TRUE);
+
+  cockpit_assert_bytes_eq (sent, "OH MARMALADE", -1);
+
+  other = g_object_ref (cockpit_peer_ensure (tc->peer));
+  emit_string (tc, NULL, "{\"command\": \"close\", \"channel\": \"a\"}");
+
+  while ((control = mock_transport_pop_control (tc->transport)) == NULL)
+    g_main_context_iteration (NULL, TRUE);
+  cockpit_assert_json_eq (control, "{\"command\":\"close\",\"channel\":\"a\"}");
+  control = NULL;
+
+  g_signal_connect (other, "closed", G_CALLBACK (on_other_closed), &closed);
+  while (!closed)
+    g_main_context_iteration (NULL, TRUE);
+
+  /* Sending again reopens with count at zero */
+  emit_string (tc, NULL, "{\"command\": \"open\", \"channel\": \"a\", \"payload\": \"upper\"}");
+  emit_string (tc, "a", "Oh MarmaLade");
+
+  while ((control = mock_transport_pop_control (tc->transport)) == NULL)
+    g_main_context_iteration (NULL, TRUE);
+  cockpit_assert_json_eq (control, "{\"command\":\"ready\",\"channel\":\"a\",\"count\":0}");
+  control = NULL;
+
+  while ((sent = mock_transport_pop_channel (tc->transport, "a")) == NULL)
+    g_main_context_iteration (NULL, TRUE);
+
+  /* The fallback just echos */
+  cockpit_assert_bytes_eq (sent, "OH MARMALADE", -1);
```

Comment:
The test case is named "timeout", but I don't see anything that resembles a timeout error or a handling thereof? Obviously there is a fallback bridge for "upper" that kicks in, but I don't see which part here waits a second for the "upper" bridge to time out? Isn't line 533/534 still talking to the originally requested "upper" bridge? This might need some more comments to understand it.